cvmfs/catalog_mgr_rw.cc

```diff
@@ -172,29 +172,34 @@ manifest::Manifest *WritableCatalogManager::CreateRepository(
 
 
 /**
- * Tries to retrieve the catalog containing the given path
- * This method is just a wrapper around the FindCatalog method of
- * AbstractCatalogManager to provide a direct interface returning
- * WritableCatalog classes.
- * @param path the path to look for
- * @param result the retrieved catalog (as a pointer)
- * @return true if catalog was found, false otherwise
+ * Retrieve the catalog containing the given path.
+ * Other than AbstractCatalogManager::FindCatalog() this mounts nested
+ * catalogs if necessary and returns  WritableCatalog objects.
+ * Furthermore it optionally returns the looked-up DirectoryEntry.
+ *
+ * @param path    the path to look for
+ * @param result  the retrieved catalog (as a pointer)
+ * @param dirent  is set to looked up DirectoryEntry for 'path' if non-NULL
+ * @return        true if catalog was found
  */
-bool WritableCatalogManager::FindCatalog(const string &path,
-                                         WritableCatalog **result)
-{
+bool WritableCatalogManager::FindCatalog(const string     &path,
+                                         WritableCatalog **result,
+                                         DirectoryEntry   *dirent) {
+  const PathString ps_path(path);
+
   Catalog *best_fit =
-    AbstractCatalogManager<Catalog>::FindCatalog(PathString(path.data(),
-                                                            path.length()));
+    AbstractCatalogManager<Catalog>::FindCatalog(ps_path);
   assert(best_fit != NULL);
   Catalog *catalog = NULL;
-  bool retval = MountSubtree(PathString(path.data(), path.length()),
-                             best_fit, &catalog);
+  bool retval = MountSubtree(ps_path, best_fit, &catalog);
   if (!retval)
     return false;
 
   catalog::DirectoryEntry dummy;
-  bool found = LookupPath(path, kLookupSole, &dummy);
+  if (NULL == dirent) {
+    dirent = &dummy;
+  }
+  bool found = catalog->LookupPath(ps_path, dirent);
```

Comment:
At this point `catalog` is guaranteed to be the correct `Catalog`. Therefore we can directly call `catalog->LookupPath()` instead of `CatalogMgr<>::LookupPath()` which would do another costly `AbstractCatalogManager::FindCatalog()`.